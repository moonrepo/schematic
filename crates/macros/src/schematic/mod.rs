use crate::common::Macro;
use crate::utils::extract_comment;
use proc_macro2::TokenStream;
use quote::{quote, ToTokens};

pub struct SchematicMacro<'l>(pub Macro<'l>);

impl<'l> ToTokens for SchematicMacro<'l> {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        let cfg = &self.0;
        let name = cfg.name;
        let schema = cfg.type_of.generate_schema(
            name,
            extract_comment(&cfg.attrs),
            cfg.get_casing_format(),
            cfg.get_tagged_format(),
        );

        tokens.extend(quote! {
            #[automatically_derived]
            impl schematic::Schematic for #name {
                fn generate_schema() -> schematic::SchemaType {
                    use schematic::schema::*;
                    #schema
                }
            }
        });
    }
}
