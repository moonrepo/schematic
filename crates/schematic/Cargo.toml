[package]
name = "schematic"
version = "0.15.1"
edition = "2021"
license = "MIT"
description = "A layered serde configuration and schema library."
homepage = "https://moonrepo.github.io/schematic"
repository = "https://github.com/moonrepo/schematic"
readme = "../../README.md"

[package.metadata.release]
pre-release-replacements = [
	{ file = "../../CHANGELOG.md", search = "Unreleased", replace = "{{version}}" },
]

[package.metadata.docs.rs]
all-features = true

[dependencies]
schematic_macros = { version = "0.15.0", path = "../macros" }
schematic_types = { version = "0.7.0", path = "../types" }
miette = { workspace = true }
thiserror = "1.0.59"
tracing = "0.1.40"

# config
garde = { version = "0.18.0", default-features = false, optional = true, features = [
	"regex",
] }
serde = { workspace = true }
serde_path_to_error = { version = "0.1.16", optional = true }
starbase_styles = { version = "0.4.0", optional = true }

# schema
indexmap = { workspace = true, optional = true, features = ["serde"] }

# json
serde_json = { workspace = true, optional = true, features = [
	"preserve_order",
] }

# json schema
markdown = { version = "1.0.0-alpha.17", optional = true }
schemars = { version = "0.8.17", optional = true, default-features = false }

# toml
toml = { workspace = true, optional = true }

# yaml
serde_yaml = { workspace = true, optional = true }

# url
reqwest = { workspace = true, optional = true, features = ["blocking"] }

[features]
default = ["config", "url"]
config = [
	"dep:garde",
	"dep:serde_path_to_error",
	"dep:starbase_styles",
	"schematic_macros/config",
]
json = ["dep:serde_json"]
schema = ["dep:indexmap", "schematic_macros/schema"]
toml = ["dep:toml"]
url = ["dep:reqwest"]
yaml = ["dep:serde_yaml"]

renderer_json_schema = ["dep:markdown", "dep:schemars", "json", "schema"]
renderer_template = []
renderer_typescript = ["schema"]

type_chrono = ["schematic_types/chrono"]
type_indexmap = ["schematic_types/indexmap"]
type_regex = ["schematic_types/regex"]
type_relative_path = ["schematic_types/relative_path"]
type_rust_decimal = ["schematic_types/rust_decimal"]
type_semver = ["schematic_types/semver"]
type_serde_json = ["schematic_types/serde_json"]
type_serde_toml = ["schematic_types/serde_toml"]
type_serde_yaml = ["schematic_types/serde_yaml"]
type_url = ["schematic_types/url"]

valid_email = ["garde/email"]
valid_url = ["garde/url"]

[dev-dependencies]
schematic = { path = ".", features = [
	"config",
	"json",
	"schema",
	"toml",
	"renderer_json_schema",
	"renderer_template",
	"renderer_typescript",
	"type_chrono",
	"type_indexmap",
	"type_regex",
	"type_relative_path",
	"type_rust_decimal",
	"type_semver",
	"type_serde_json",
	"type_serde_toml",
	"type_serde_yaml",
	"type_url",
	"url",
	"valid_email",
	"valid_url",
	"yaml",
] }
reqwest = { workspace = true, features = [
	"blocking",
	"rustls-tls-native-roots",
] }
serial_test = "3.1.1"
starbase_sandbox = "0.6.0"

# Types
chrono = { workspace = true }
indexmap = { workspace = true }
regex = { workspace = true }
relative-path = { workspace = true, features = ["serde"] }
rust_decimal = { workspace = true }
semver = { workspace = true, features = ["serde"] }
url = { workspace = true, features = ["serde"] }
