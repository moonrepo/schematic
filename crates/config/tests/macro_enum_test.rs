use schematic::*;

// #[derive(Config)]
// enum AllUnit {
//     Foo,
//     Bar,
//     Baz,
// }

// #[derive(Config)]
// enum AllUnnamed {
//     Foo(String),
//     Bar(bool),
//     Baz(usize),
// }

// #[derive(Config)]
// enum AllNamed {
//     Foo { one: String },
//     Bar { two: bool },
//     Baz { three: usize },
// }
