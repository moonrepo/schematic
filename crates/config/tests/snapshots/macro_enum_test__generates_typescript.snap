---
source: crates/config/tests/macro_enum_test.rs
expression: "std::fs::read_to_string(file).unwrap()"
---
// Automatically generated by schematic. DO NOT MODIFY!

/* eslint-disable */

export type AllUnit = 'foo' | 'bar' | 'baz';

export type AllUnnamed = {
	foo: string;
} | {
	bar: boolean;
} | {
	baz: number;
};

export type OfBothTypes = {
	foo: null;
} | {
	bar: [boolean, number];
};

export interface SomeConfig {
	bar: number;
	foo: string;
}

export type NestedConfigs = {
	string: string;
} | {
	object: SomeConfig;
} | {
	objects: [SomeConfig, SomeConfig];
};

export type WithSerde = string | boolean | number;

export type WithComments = 'foo' | 'bar' | 'baz';

export type Untagged = null | boolean | [number, string] | SomeConfig;

export type ExternalTagged = {
	foo: 'foo';
} | {
	bar: boolean;
} | {
	bazzer: number;
} | {
	qux: SomeConfig;
};

export type InternalTagged = 'foo' | boolean | number | SomeConfig;

export type AdjacentTagged = {
	type: 'foo';
	content: 'foo';
} | {
	type: 'bar';
	content: boolean;
} | {
	type: 'bazzer';
	content: number;
} | {
	type: 'qux';
	content: SomeConfig;
};

export type PartialAllUnnamed = {
	foo: string;
} | {
	bar: boolean;
} | {
	baz: number;
};

export type PartialOfBothTypes = {
	foo: null;
} | {
	bar: [boolean, number];
};

export interface PartialSomeConfig {
	bar?: number | null;
	foo?: string | null;
}

export type PartialNestedConfigs = {
	string: string;
} | {
	object?: PartialSomeConfig | null;
} | {
	objects?: [PartialSomeConfig, PartialSomeConfig] | null;
};

export type PartialWithSerde = string | boolean | number;

export type PartialUntagged = null | boolean | [number, string] | PartialSomeConfig;

export type PartialExternalTagged = {
	foo: 'foo';
} | {
	bar: boolean;
} | {
	bazzer: number;
} | {
	qux?: PartialSomeConfig | null;
};

export type PartialInternalTagged = 'foo' | boolean | number | PartialSomeConfig;

export type PartialAdjacentTagged = {
	type: 'foo';
	content: 'foo';
} | {
	type: 'bar';
	content: boolean;
} | {
	type: 'bazzer';
	content: number;
} | {
	type?: 'qux' | null;
	content?: PartialSomeConfig | null;
};

